transcom/mymove
# MB-2942 SPIKE: Create mTLS webhook-client and support endpoint for testing

## Description

[MB-2942](https://dp3.atlassian.net/browse/MB-2942) is part of the Push Enable work. The spike was to create two independent apps that could communicate via mTLS: one to send a message, one to receive the message. 

Part of the work in the spike revealed that we can use the existing Support API/Milmove server as our testing server, since it's mTLS-enabled and already built. This PR adds a new client, `webhook-client`, that can send a message to a new support endpoint, `/support/v1/webhook-notify`. 

This is foundational work that will be refactored and built upon in future stories. It has no impact on existing functionality.

A helpful diagram created by @shimonacarvalho  is included at the bottom to illustrate the flow of the current iteration, plus planned behavior on staging once we have the appropriate certs. We determined that using a CAC won't be helpful when testing, so support for CACs won't be needed in the build-out.


## Reviewer Notes

* `cmd/webhook-client` uses the `orders-api-client` in the [milmove_orders](https://dp3.atlassian.net/browse/MB-2942) repo as its base. There's a lot of code addition, but less original code/changes that have been made.
* You can (and should!) test both via Postman and the `webhook`client`.
* `#TODO` comments have been left behind to correspond with future stories

## Setup

To test the `webhook-client`:

```sh
$ make server_run

// In a separate window,

$ make bin/webhook-client
$ webhook-client post-webhook-notify --message "Lorem ipsum" --port 9443 --hostname primelocal --insecure --certpath config/tls/devlocal-mtls.cer --keypath config/tls/devlocal-mtls.key

// Output:
{"message": "Lorem ipsum"}
```

To test in Postman:

* Send a POST request to `/support/v1/webhook-notify` with the body:
```json
{
    "message": "Lorem ipsum"
}

```
You should receive a response with your message included.

## Code Review Verification Steps

## References
* [Jira story](https://dp3.atlassian.net/browse/MB-2942) for this change

## Screenshots

<img width="985" alt="Screen Shot 2020-07-31 at 4 10 21 PM" src="https://user-images.githubusercontent.com/6007259/89087206-57a8ae80-d348-11ea-9336-dbc42f66a07f.png">
